_posts/2022-05-09-project-loom-for-distributed-systems.md

```diff
@@ -0,0 +1,397 @@
+
+## Conclusion
+
+It's typical to test the consistency protocols of distributed systems via randomized failure testing. Two approaches which sit at different ends of the spectrum are Jepsen and the simulation mechanism pioneered by FoundationDB. The former allows the system under test to be implemented in any way, but is only viable as a last line of defense. The latter can be used to guide a much more aggressive implementation strategy, but requires the system to be implemented in a very specific style.
+
+| Feature        | Jepsen-style              | FoundationDB-style            |
+| ---------------------- | --------------------------------------- | ----------------------------------------- |
+| Integration difficulty | So long as it uses syscalls, it can be done  | Required writing own programming language |
+| Speed of testing     | Slow - good for acceptance tests    | Fast - much faster than the production DB |
+| Representative?    | Yes - all subsystems same as in production | No - detailed simulations, but test doubles relied upon |
+| Useful for debugging   | No - distributed systems failures never fun to debug. | Yes - since deterministic, all failures replayable |
+
+Java's Project Loom considerably shakes up this tradeoff.
+
+![spectrum](/assets/images/spectrum.jpg){: width="100%" }
+
+# Project Loom
+
+[Project Loom](https://wiki.openjdk.java.net/display/loom/Main) provides 'virtual' threads as a first class concept within Java. There is plenty of good information in the 2020 blog post '[State of Loom](http://cr.openjdk.java.net/~rpressler/loom/loom/sol1_part1.html)' although details have changed in the last two years.
+
+I will give a simplified description of what I find exciting about this. In traditional Java we can launch a Thread. If it needs to pause for some reason, the thread will be paused, and will resume when it is able to. Java does not make it easy to control the threads (pause at a critical section, pick who acquired the lock, etc), and so influencing the interleaving of execution is very difficult except for in very isolated cases.
+
+With Loom's virtual threads, when a thread starts, a `Runnable` is submitted to an `Executor`. When that task is run by the executor, if the thread needs to block, the submitted runnable will exit, instead of pausing. When the thread can be unblocked, a new runnable is submitted to the same executor to pick up where the previous `Runnable` left off. Here, interleaving is much, much easier, since we are passed each piece of runnable work as it becomes runnable. Combined with the `Thread.yield()` primitive, we can also influence the points at which code becomes deschedulable.
+
+```java
+Queue<Runnable> executor = new ArrayDeque<>();
+Lock lock = new ReentrantLock();
+lock.lock();
+newVirtualThread(executor::add, lock::lock);
```

Comment:
it's not so clear to me how this works, because newVirtualThread() isn't explained